### PR TITLE
masterclass: Initial PR for services masterclass

### PR DESCRIPTION
**Pull request by hedss**
Change-type: major
Signed-off-by: Heds Simons <heds@balena.io>

### CODE REVIEW

File: multi-stage/Dockerfile.template [hunk 1]
```diff
@@ -0,0 +1,14 @@
+FROM balenalib/%%BALENA_MACHINE_NAME%%-debian:buster
+
+# Install build essential tools to include GCC
+RUN apt-get update && apt-get install -y \
```
Comment (shaunmulligan) [inline multi-stage/Dockerfile.template]: is it not better to use the `RUN install_packages` which will do the apt update, retries and clean up as well

File: multicontainer-app/backend/Dockerfile.template [hunk 2]
```diff
@@ -0,0 +1,11 @@
+FROM balenalib/%%BALENA_MACHINE_NAME%%-node:10
+
+WORKDIR /usr/src/app
+
+COPY package* ./
+
+RUN npm ci --production && npm cache clean --force && rm -rf /tmp/*
```
Comment (shaunmulligan) [inline multicontainer-app/backend/Dockerfile.template:7]: I think it we want this to work well with localMode we might not want to optimize with the cache cleaning, etc?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,1584 @@
-# services-masterclass
-A guide to developing and deploying services on a balena device
+balena Services Masterclass
+===========================
+
+A guide to developing and deploying services on a balena device.
+
+**Masterclass Type:** Core
+**Maximum Expected Time To Complete:**
```
Comment (shaunmulligan) [inline README.md:13]: I guess we don't yet have a time estimate on this one?

File: README.md [hunk 4]
```diff
@@ -1,2 +1,1584 @@
...
+
+* A balena supported device, such as a [balenaFin 1.1](https://store.balena.io/collections/developer-kit/products/balenafin-v1-1-0-developer-kit),
+        [Raspberry Pi 3](https://www.raspberrypi.org/products/raspberry-pi-3-model-b/)
+        or [Intel NUC](https://www.intel.co.uk/content/www/uk/en/products/boards-kits/nuc.html). If you don't have a device, you can emulate an Intel NUC by
+        installing VirtualBox and following [this guide](https://www.balena.io/blog/no-hardware-use-virtualbox/)
+* A suitable text editor for developing code on your development platform (eg.
+    [Visual Code](https://code.visualstudio.com/))
+* A suitable shell environment for command execution (such as `bash`)
+* A [balenaCloud](https://www.balena.io/) account
+* Some familiarity with [Dockerfiles](https://docs.docker.com/engine/reference/builder/) and [docker-compose]()
```
Comment (shaunmulligan) [inline README.md]: missing link to docker-compose, probably want version 2 docs: https://docs.docker.com/compose/compose-file/compose-file-v2/

File: README.md [hunk 5]
```diff
@@ -1,2 +1,1584 @@
...
+* An installation of [balena CLI](https://github.com/balena-io/balena-cli/blob/master/INSTALL.md)
+
+# Exercises
+
+All of the following exercises assume that you are able to access a suitable
+Unix based shell and that balena CLI is installed. The exercises include
```
Comment (shaunmulligan) [inline README.md:45]: does this class assume having done the CLI master class first?

File: README.md [hunk 6]
```diff
@@ -1,2 +1,1584 @@
...
+
+## 1.1 Service Images
+
+Service images are the basis for any service. They comprise the binaries and
+data for running services, but they also contain metadata that informs
+balenaEngine on how to initialise and configure the service. As such, an image
+is essentially a self-enclosed Linux filesystem with everything required for
+running an instance of Linux, except for the kernel.
+
+Images are produced by the balena builders, which use Dockerfiles to determine
```
Comment (shaunmulligan) [inline README.md]: they could of course be built anywhere else (localmode or just as `image:` reference in a compose. but i think its probably best not to get to into the weeds here.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,1584 @@
...
+
+Volumes are persistent storage areas which do not get erased when a service
+container is recreated. These volumes are usually bound to a service container
+on a specific path, and are stored on the host filesystem without any
+sort of diff mechanism (that is, any file hierarchy data written to the volume
+is always available in that hierarchy).
+
+For single service applications, there is a default persistent volume bound to
+the `/data` directory in the service. This path can always be considered
+persistent.
```
Comment (shaunmulligan) [inline README.md:124]: in the single container case only. we have had people expect it to also to be there in multicontainer and been confused when data is not persisted.

**Review by shaunmulligan — COMMENTED** [review]
still working through the review, but first set of comments.

File: multicontainer-app/backend/Dockerfile.template (hunk 2 above)
Comment (hedss) [reply multicontainer-app/backend/Dockerfile.template:7]: We could, but I don't actually use local mode in this masterclass.

File: README.md (hunk 3 above)
Comment (hedss) [reply README.md:13]: Not yet, I'm hoping someone will go through it. Hint hint. :)

File: README.md (hunk 4 above)
Comment (hedss) [reply README.md]: Yep, good spot.

File: README.md (hunk 5 above)
Comment (hedss) [reply README.md:45]: Yeah, we should say that.

File: README.md (hunk 6 above)
Comment (hedss) [reply README.md]: Yeah, have clarified potential sources.

File: multi-stage/Dockerfile.template (hunk 1 above)
Comment (hedss) [reply multi-stage/Dockerfile.template]: Have changed.

File: README.md [hunk 8]
```diff
@@ -1,2 +1,1610 @@
...
+the following masterclasses:
+
+* [balena CLI Masterclass](https://github.com/balena-io-projects/balena-cli-masterclass)
+
+**Masterclass Type:** Core
+**Maximum Expected Time To Complete:**
+
+# Introduction
+
+balena applications consist of one or more balena services, which are based upon
```
Comment (xginn8) [inline README.md]:
```suggestion
Balena applications consist of one or more balena services, which are based upon
```

File: README.md [hunk 9]
```diff
@@ -1,2 +1,1610 @@
...
+balenaEngine (a fork of Docker) images, containers, volumes and networks. This
```
Comment (xginn8) [inline README.md]:
```suggestion
Docker images, containers, volumes and networks. This
```

File: README.md [hunk 10]
```diff
@@ -1,2 +1,1610 @@
...
+appended under the line to show what might be returned. For example:
+
+```
+$ balena version
+11.11.2
+```
+
+# 1. Service Images, Containers, Volumes and Networks
+
+balena applications are comprised of service images, containers and volumes.
```
Comment (xginn8) [inline README.md]:
```suggestion
Balena applications are comprised of service images, containers and volumes.
```

File: README.md [hunk 11]
```diff
@@ -1,2 +1,1610 @@
...
+from any directory within that service on these locally created filesystem
+layers. However, these layers only exist for the lifespan of the container.
+Containers are recreated for multiple reasons. These can include newer versions
+of a service image being downloaded, alterations to environment variables names
+or values, a change in network configuration, etc.
+
+As such, it is never safe to rely on data existing in a service container layer
+above that of the lifespan of that service container. Temporary data is
+therefore acceptable use, but any data that is required to persist across
+container recreation or restarts should be stored in persistent volumes.
```
Comment (xginn8) [inline README.md]: restarts _should_ be safe for local data, right? (though i know https://github.com/balena-io/balena-supervisor/issues/1116 makes that currently untrue)

File: README.md [hunk 12]
```diff
@@ -1,2 +1,1610 @@
...
+Both types of applications are detailed in the following exercises.
+
+# 2. Single Service Applications
+
+Single service applications, as the name suggests, consist of a single unique
+service. These are defined using a `Dockerfile`, `Dockerfile.template` or
+`package.json` file, which is built either by a balena environment, a host
+development machine or a balena device. The resulting service image is then
+downloaded to all devices provisioned against their application and executed
+on-device by the balena Supervisor.
```
Comment (xginn8) [inline README.md]: small distinction: on reboot, this container is actually started by balenaEngine (not the super directly) iiuic

File: README.md [hunk 13]
```diff
@@ -1,2 +1,1610 @@
...
+paradigm) are required for either privilege level or to ensure self-contained
+networking, then a multicontainer application should be used (even if only one
+service is required).
+
+## 2.1 Building and Deploying a Single Service Application
+
+Change directory to the `single-service-app` directory in the root of this
+masterclass repository, eg:
+```
+$ cd ~/services-masterclass/single-service-app
```
Comment (xginn8) [inline README.md]: meta: it _might_ be a tiny bit easier to just use an envvar here (ie i don't like that path, but i can just export my own `$BALENA_MASTERCLASSES_PATH`)

File: README.md [hunk 14]
```diff
@@ -1,2 +1,1610 @@
...
+the application.
+
+Unlike single service applications, balenaEngine will generate an internal
+bridge network for services to communicate between themselves. This bridge
+allows the ingress of network traffic from host interfaces only to services
+which have explicitly specified that they listen on given ports.
+
+## 3.1 Multicontainer `docker-compose` Manifest
+
+balena allows the configuration of a multicontainer service application via the
```
Comment (xginn8) [inline README.md]:
```suggestion
Balena allows the configuration of a multicontainer service application via the
```

File: README.md [hunk 15]
```diff
@@ -1,2 +1,1610 @@
...
+any capabilities, permissions, access, environment variables, etc. that each
+service may require, as well as the definition and configuration of any shared
+resources that those services may require. One very important difference between
+single service and multicontainer applications is that multicontainer services
+are *not* privileged by default. This means that if you want access to host
+device nodes, capabilities, etc. then you need to either enable privileged
+access within the `docker-compose` manifest, or pick a subset of the access
+that the service requires using relevant keywords in that service definition.
+
+To demonstrate a multicontainer application, we'll start with a  couple of very
```
Comment (xginn8) [inline README.md]:
```suggestion
To demonstrate a multicontainer application, we'll start with a couple of very
```

File: README.md [hunk 16]
```diff
@@ -1,2 +1,1610 @@
...
+* `bridge` mode uses a bridge network interface which all service containers
+    are connected to
+
+This means a couple of things when developing applications. Any service that
+uses `host` networking does not have to explicitly define ports for traffic
+ingress, and a service can bind to all interface and expect incoming traffic
+to the host. Single service applications *always* use `host` networking.
+
+By contrast, `bridge` networking essentially isolates all services from the
+host, requring services to explicitly define open ports to allow traffic from
```
Comment (xginn8) [inline README.md]:
```suggestion
host, requiring services to explicitly define open ports to allow traffic from
```

File: README.md [hunk 17]
```diff
@@ -1,2 +1,1610 @@
...
+the host to be passed to them (although all outgoing connections are allowed
+and will allow incoming data from those connections). By default, multicontainer
+application serviers always use a self-defined `bridge` network.
```
Comment (xginn8) [inline README.md]:
```suggestion
application servers always use a self-defined `bridge` network.
```

File: README.md [hunk 18]
```diff
@@ -1,2 +1,1610 @@
...
+
+Given this knowledge, and how we've defined our `frontend` and `backend`
+services, you can see that we're using both `host` (for the `frontend` service)
+and `bridge` (for the `backend` service) networking. The backend shouldn't
+really be exposed, as we want the frontend to retrieve data from it and
+then pass it to an external query.
+
+Here's a diagram showing what's currently happens with network traffic:
+
+<Diagram where connections can occur both to frontend and backend>
```
Comment (xginn8) [inline README.md]: missing diagram?

File: README.md [hunk 19]
```diff
@@ -1,2 +1,1610 @@
...
+
+And here's a diagram showing what we want to happen:
+
+<Diagram where frontend only is exposed, and the backend can be communicated with
```
Comment (xginn8) [inline README.md]: ditto

File: README.md [hunk 20]
```diff
@@ -1,2 +1,1610 @@
...
+    display-manager.service \
+    getty@.service \
+    systemd-logind.service \
+    systemd-remount-fs.service \
+    getty.target \
+    graphical.target
+```
+This installs `systemd` as well as `dbus` (which `systemd` requires) together
+into a suitable Debian base image and then masks services that do should not be
+run inside a container:
```
Comment (xginn8) [inline README.md]:
```suggestion
run inside a container.
```

perhaps?

File: README.md [hunk 21]
```diff
@@ -1,2 +1,1610 @@
...
+act as the init process. We'll copy this entry script into the service image
+and then set it as the service entrypoint. Add the following to then end of
+what you've already added to `services-masterclass/printer/Dockerfile.template`:
+```
+# Copy our systemd entrypoint script and ensure it's used
+COPY entry.sh /usr/bin
+STOPSIGNAL 37
+ENTRYPOINT ["/usr/bin/entry.sh"]
+```
+Note the `STOPSIGNAL` directive. By default balenaEngine (and Docker) will send
```
Comment (xginn8) [inline README.md:745]: nice, didn't know this, TIL!

File: README.md [hunk 22]
```diff
@@ -1,2 +1,1610 @@
...
+			  \`,      `-.`7  7)   )
+			   \/         \|  \'  / `-._
+			              ||    .'
+			               \\  (
+			                >\  >
+			            ,.-' >.'
+			           <.'_.''
+			             <'
+```
+We've cut down the build log a bit, to make it shorted. Wait for the application
```
Comment (xginn8) [inline README.md]:
```suggestion
We've cut down the build log a bit, to make it shorter. Wait for the application
```

File: README.md [hunk 23]
```diff
@@ -1,2 +1,1610 @@
...
+etc-hosts.mount                    loaded active     mounted   /etc/hosts
+etc-resolv.conf.mount              loaded active     mounted   /etc/resolv.conf
+sys-kernel-debug.mount             loaded active     mounted   Kernel Debug File System
+...
+```
+
+Whilst we've now mostly dealt with running `systemd` in a container, for
+completeness we'll also show how to quickly add a service so it starts when the
+service container is started. Add the following to the end of the
+``services-masterclass/printer/Dockerfile.template` file:
```
Comment (xginn8) [inline README.md]:
```suggestion
`services-masterclass/printer/Dockerfile.template` file:
```

File: README.md [hunk 24]
```diff
@@ -1,2 +1,1610 @@
...
+```
+
+Ensure you've created a new named application (for example, 'Multistage')
+for your balenaFin (or alternative device). Download and provision a development
+device image for that application and flash it onto your device. If you haven't
+already done so, login to your account using balena CLI with `balena login` and
+choose an appropriate authentication method. Make a note of the UUID of the
+device.
+
+To demonstrate mulitstage builds, look at the
```
Comment (xginn8) [inline README.md]:
```suggestion
To demonstrate multistage builds, look at the
```

File: README.md [hunk 25]
```diff
@@ -1,2 +1,1610 @@
...
+[main]     Selecting previously unselected package libubsan1:armhf.
+[main]     Preparing to unpack .../18-libubsan1_8.3.0-6+rpi1_armhf.deb ...
+[main]     Unpacking libubsan1:armhf (8.3.0-6+rpi1) ...
+[main]     Selecting previously unselected package libgcc-8-dev:armhf.
+[main]     Preparing to unpack .../19-libgcc-8-dev_8.3.0-6+rpi1_armhf.deb ...
+[main]     Unpacking libgcc-8-dev:armhf (8.3.0-6+rpi1) ...
+[main]     Selecting previously unselected package gcc-8.
+[main]     Preparing to unpack .../20-gcc-8_8.3.0-6+rpi1_armhf.deb ...
+[main]     Unpacking gcc-8 (8.3.0-6+rpi1) ...
+[main]     Selec
```
Comment (xginn8) [inline README.md:1590]:
```suggestion
Now our service is only 39.9MB in size, as opposed to the 289MB it was
```

Comment (xginn8) [inline README.md]: can be removed, i suppose

**Review by xginn8 — APPROVED** [review]

File: README.md (hunk 11 above)
Comment (hedss) [reply README.md]: Yeah, I've removed restarts but will keep the info in about the issue in the Supervisor.

File: README.md (hunk 12 above)
Comment (hedss) [reply README.md]: Yes, good point.

File: README.md (hunk 13 above)
Comment (hedss) [reply README.md]: Yeah, they still need to set the var but at least it can be used in the rest.

File: README.md (hunk 14 above)
Comment (hedss) [reply README.md]: Done

File: README.md (hunk 10 above)
Comment (hedss) [reply README.md]: Done

File: README.md (hunk 9 above)
Comment (hedss) [reply README.md]: Done

File: README.md (hunk 8 above)
Comment (hedss) [reply README.md]: Done

File: README.md (hunk 16 above)
Comment (hedss) [reply README.md]: Done

File: README.md (hunk 15 above)
Comment (hedss) [reply README.md]: Done

File: README.md (hunk 17 above)
Comment (hedss) [reply README.md]: Done

File: README.md (hunk 18 above)
Comment (hedss) [reply README.md]: Yup!

File: README.md (hunk 19 above)
Comment (hedss) [reply README.md]: Yup!

File: README.md (hunk 20 above)
Comment (hedss) [reply README.md]: Done

File: README.md (hunk 22 above)
Comment (hedss) [reply README.md]: Done

File: README.md (hunk 24 above)
Comment (hedss) [reply README.md]: Done

File: README.md (hunk 23 above)
Comment (hedss) [reply README.md]: Done

File: README.md (hunk 25 above)
Comment (hedss) [reply README.md]: Have removed references.

File: README.md [hunk 26]
```diff
@@ -1,2 +1,1613 @@
...
+[main]     Selecting previously unselected package libasan5:armhf.
+[main]     Preparing to unpack .../17-libasan5_8.3.0-6+rpi1_armhf.deb ...
+[main]     Unpacking libasan5:armhf (8.3.0-6+rpi1) ...
+[main]     Selecting previously unselected package libubsan1:armhf.
+[main]     Preparing to unpack .../18-libubsan1_8.3.0-6+rpi1_armhf.deb ...
+[main]     Unpacking libubsan1:armhf (8.3.0-6+rpi1) ...
+[main]     Selecting previously unselected package libgcc-8-dev:armhf.
+[main]     Preparing to unpack .../19-libgcc-8-dev_8.3.0-6+rpi1_armhf.deb ...
+[main]     Unpacking libgcc-8-dev:armhf (8.3.0-6+rpi1) ...
+[main]     Selecting previou
```
Comment (xginn8) [inline README.md:1598]:
```suggestion
[multi-stage builds here](https://docs.docker.com/develop/develop-images/multistage-build/).
```

**Pull request merged**